sep/sep-022.rst

```diff
@@ -0,0 +1,119 @@
+=======  ==================
+SEP      22
+Title    Procssors Overhaul
+Author   Mikhail Korobov
+Created  2015-02-10
+Status   Draft
+=======  ==================
+
+===========================
+SEP-022: Procssors Overhaul
+===========================
+
+There is scrapy.contrib.loader and scrapy.contrib.loader.processor modules
+which allow to create declarative pipelines for data processing and use these
+pipelines to populate Scrapy Items.
+
+Processors Overview
+===================
+
+Let's start with Compose and TakeFirst.
+
+**Compose** - returns a combined function which calls all its functions in order.
+**TakeFirst** - convert from a sequence to a single item by taking first non-empty value.
+
+Analogues of ``func = Compose(TakeFirst(), unicode.strip)``:
+
+* no libraries:
+
+  >>> func = lambda it: it[0].strip()  # almost the same, but not composable
+  >>> func = lambda it: unicode.strip(TakeFirst()(it))
+
+* http://toolz.readthedocs.org/en/latest/api.html#toolz.functoolz.compose
+  (applies functions in a different order):
+
+  >>> from toolz import compose, first
+  >>> func = compose(unicode.strip, TakeFirst())
+  >>> func = compose(unicode.strip, first)
+
+* http://funcy.readthedocs.org/en/stable/funcs.html#compose
+  (applies functions in a different order);
+
+  >>> from funcy import compose, first
+  >>> func = compose(unicode.strip, TakeFirst())
+  >>> func = compose(unicode.strip, first)
+
+* https://github.com/kachayev/fn.py provides F object which allows to
+  compose functions; syntax is a bit unusual:
+
+  >>> from fn import F
+  >>> from fn.iters import first
+  >>> func = F() >> TakeFirst() >> unicode.strip
+  >>> func = F() >> first >> unicode.strip        # the same
+  >>> func = F() << unicode.strip << TakeFirst()  # the same
+
+
+Notes:
+
+1. Compose has support for ``loader_context``. I personally have never
```

Comment:
These are some examples I found:
```python
def add_base(loader, values, loader_context=None):
    response = loader_context['response']
    return [urljoin(get_base_url(response), ''.join(x)) for x in values]
```
```python
# eval contains stuff like
def nl_any(values):
    if values:
        return any(_nl_yes_re.search(s) for s in values)

def contextual_eval(values, loader_context):
    if values:
        return loader_context['eval'](values)
```
```python
def populate_name(value, loader_context):
    item = loader_context['item']
    parts = value.split()
    # major case: name is two words only.
    # Otherwise will need smarter heuristics to tell which part is the last
    # name. For example: "R Van Berkel", "Minaz H. K. Rattansi", "Vi Ha Nguyen"
    if len(parts) == 2:
        item['first_name'] = parts[0]
        item['last_name'] = parts[1]
    return value
```
```python
def join_base_url(url, loader_context):
    return urljoin(loader_context['base_url'], url)
```
```python
def transform_into_btc(dummy, loader_context):
    price = float_safe_cast(loader_context.get('usd_price'))
    if price is not None:
        return BitcoinConversionAPI.get_from_usd(price)
    price = float_safe_cast(loader_context.get('eur_price'))
    if price is not None:
        return BitcoinConversionAPI.get_from_eur(price)
    return None
```
```python
def parse_sold_count(text, loader_context):
    try:
        text = text.replace(",", "").replace(".", "")
        n = parse_int(text)
        return n
    except Exception as e:
        e2 = SpiderMissingValueException(spider_name=loader_context['spider_name'], field_name=FIELD_SOLD_COUNT, url=loader_context[FIELD_URL], exception=e, level=STATUS_CRITICAL)
        log_alert(e2, loader_context['spider_name'] + '_' + FIELD_SOLD_COUNT)
        return None
```